### PR TITLE
[MODINREACH-169] INN-Reach Circulation Flow: React to Loan Record Changes for INN-Reac…

### PR DESCRIPTION
**Pull request by aleksandr2110**
**Purpose**
1. mod-inn-reach should consume loan-storage domain events via Kafka (CIRCSTORE-300)
2. mod-inn-reach should check if the loan is associated with an active INN-Reach transaction
3. If so, and the last loan action is "checkedin", the loan status is "Closed" and the INN-Reach transaction is a an Item Hold, remove the due date, patron name, and patron ID from the transaction record,  issue the final check-in message to the central server (D2IR API Reference pg. 54), and update the transaction status to FINAL_CHECKIN.

**Approach:**
Check-in of item that is loaned to a borrowing site reliably triggers a final check-in message to the central server and related transaction status update to FINAL_CHECKIN for the INN-Reach transaction
The patron name and patron ID values are also removed from the transaction

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/resources/db/changelog/changes/v1.1.0/sql/2022-01-28-01__alter-transaction-hold-tables.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE transaction_hold DROP COLUMN patron_id;
+ALTER TABLE transaction_hold ADD COLUMN patron_id VARCHAR(32) CHECK(patron_id SIMILAR TO '[a-z,0-9]{1,32}');
+ALTER TABLE transaction_item_hold DROP COLUMN patron_name;
+ALTER TABLE transaction_item_hold ADD COLUMN patron_name VARCHAR(255);
```
Comment (oliinyko) [inline src/main/resources/db/changelog/changes/v1.1.0/sql/2022-01-28-01__alter-transaction-hold-tables.sql]: alter existing columns

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java [hunk 2]
```diff
@@ -33,4 +34,7 @@
 
   @Query(name = FETCH_OPEN_BY_LOAN_ID_QUERY_NAME)
   Optional<InnReachTransaction> fetchOneByLoanId(UUID folioLoanId);
+
+  @Query(name = FETCH_BY_LOAN_ID_QUERY_NAME)
+  Optional<InnReachTransaction> fetchTransactionByLoanId(UUID folioLoanId);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: why `fetchOneByLoanId` is not used instead?

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: They are different

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 3]
```diff
@@ -129,30 +134,46 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
 
   @Override
   public void handleLoanUpdate(LoanDTO loan) {
+    InnReachTransaction associatedTransactionWithLoan = null;
+    Optional<InnReachTransaction> optionalTransaction = transactionRepository.fetchOneByLoanId(loan.getId());
+    if (optionalTransaction.isPresent()) {
+      associatedTransactionWithLoan = optionalTransaction.get();
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: `return` if not present

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 4]
```diff
@@ -129,30 +134,46 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
...
+        reportFinalCheckIn(associatedTransactionWithLoan);
+        transactionRepository.save(associatedTransactionWithLoan);
+      }
+      else {
+        updateAssociatedTransaction(loan, transaction -> {
+          log.info("Updating transaction {} on loan closure {}", transaction.getId(), loan.getId());
+          transaction.getHold().setDueDateTime(null);
+          transaction.setState(ITEM_IN_TRANSIT);
+          reportItemInTransit(transaction);
+        });
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: remove this call `updateAssociatedTransaction` and reuse `associatedTransactionWithLoan`

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 5]
```diff
@@ -129,30 +134,46 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
 
   @Override
   public void handleLoanUpdate(LoanDTO loan) {
+    InnReachTransaction associatedTransactionWithLoan = null;
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: rename it to `transaction`, it should be obvious from the context that the transaction is associated with loan

Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: use `var` consistently

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: var doesn't work here

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:161]: Done

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 3 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Done

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 2 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: It's obvious that they are different, the question is why different query is needed here. 
Pay attention to the requirements:
> mod-inn-reach should check if the loan is associated with **_an active_** INN-Reach transaction

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 5 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]:
```
    var transaction = transactionRepository.fetchOneByLoanId(loan.getId()).orElse(null);
    if (transaction == null) {
      return;
    }
```

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:137]: I am going to use  if (optionalTransaction.isPresent()) {
      transaction = optionalTransaction.get();
    } else {
      new EntityNotFoundException("INN-Reach transaction is not found by loan id: " + loan.getId());
    } isn't ok ?

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Got it

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 6]
```diff
@@ -186,6 +188,51 @@ void shouldRenewalLoanToUpdateTransaction() {
...
+
+    var capturedEvent = capturedEvents.get(0);
+    assertEquals(folioLoanId, capturedEvent.getRecordId());;
+    Optional<InnReachTransaction> optionalTransaction = transactionRepository.fetchTransactionByLoanId(folioLoanId);
+    InnReachTransaction transaction = null;
+    if (optionalTransaction.isPresent()) {
+      transaction = optionalTransaction.get();
+    }
+    var transactionHold = (TransactionHold) transaction.getHold();
+    var transactionItemHold = (TransactionItemHold) transaction.getHold();
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: `var transactionHold = (TransactionItemHold) transaction.getHold();`

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 2 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: This query uses for testing or better not to use such query ?

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 6 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Got it

File: src/main/resources/db/changelog/changes/v1.1.0/sql/2022-01-28-01__alter-transaction-hold-tables.sql (hunk 1 above)
Comment (aleksandr2110) [reply src/main/resources/db/changelog/changes/v1.1.0/sql/2022-01-28-01__alter-transaction-hold-tables.sql]: Done

File: src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java (hunk 2 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/repository/InnReachTransactionRepository.java]: Resolved

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 7]
```diff
@@ -129,30 +134,40 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
...
-      updateAssociatedTransaction(loan, transaction -> {
-        log.info("Updating transaction {} on loan closure {}", transaction.getId(), loan.getId());
+      if (transaction != null && (transaction.getType() == ITEM)) {
+        log.info("Updating transaction {} on loan to final check-in {}", transaction.getId(), loan.getId());
         transaction.getHold().setDueDateTime(null);
-        transaction.setState(ITEM_IN_TRANSIT);
-        reportItemInTransit(transaction);
-      });
+        var transactionHold = (TransactionHold) transaction.getHold();
+        var transactionItemHold = (TransactionItemHold) transaction.getHold();
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: why do you need both?

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 4 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 8]
```diff
@@ -129,30 +132,41 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
 
   @Override
   public void handleLoanUpdate(LoanDTO loan) {
+    var associateNewLoanWithTransaction = transactionRepository.fetchOneByLoanId(loan.getId()).orElse(null);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: https://github.com/folio-org/mod-inn-reach/pull/181#discussion_r795985907

Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Done

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 9]
```diff
@@ -129,30 +132,41 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
...
-        var loanDueDate = loan.getDueDate().toInstant().truncatedTo(ChronoUnit.SECONDS);
-        if (!loanDueDate.equals(transactionDueDate)) {
-          var loanIntegerDueDate = (int) (loanDueDate.getEpochSecond());
-          reportBorrowerRenew(transaction, loanIntegerDueDate);
-          transaction.setState(BORROWER_RENEW);
-          transaction.getHold().setDueDateTime(loanIntegerDueDate);
-          transactionRepository.save(transaction);
-        }
-      });
+    if (loan.getAction().equals("renewed") && transaction != null) {
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: `transaction != null` is redundant

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 10]
```diff
@@ -119,13 +120,16 @@ void shouldLinkLoanToOpenTransaction() {
...
   })
   void shouldReturnCheckedInItem() {
+    UUID folioLoanId = UUID.fromString("fd5109c7-8934-4294-9504-c1a4a4f07c96");
     var event = getLoanDomainEvent(DomainEventType.UPDATED);
     event.setRecordId(null); // the listener should set this field from event key value
     var loan = event.getData().getNewEntity();
+    loan.setId(folioLoanId);
     loan.setStatus(new LoanStatus().name("Closed"));
     loan.setAction("checkedin");
+    event.setData(new EntityChangedData<>(null, loan));
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: redundant call

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 11]
```diff
@@ -119,13 +120,16 @@ void shouldLinkLoanToOpenTransaction() {
     "classpath:db/inn-reach-transaction/pre-populate-inn-reach-transaction.sql"
   })
   void shouldReturnCheckedInItem() {
+    UUID folioLoanId = UUID.fromString("fd5109c7-8934-4294-9504-c1a4a4f07c96");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: use `var` consistently

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 9 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 11 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 10 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Deleted

**Review by oliinyko — APPROVED** [review]

**Pull request merged**